### PR TITLE
frontier config

### PR DESCRIPTION
**Pull request by oksanaguba**
For start, used Andrew's tag [screamv1-frontier-feb2023](https://github.com/E3SM-Project/scream/tree/archive/screamv1-frontier-feb2023) and merged master into it.

Changed account, FS, changed CC compiler to get correct linking flags for threaded runs for land blas calls (but it resolved crashes in land around blas calls only partially).

[bfb]

### CODE REVIEW

**Comment by oksanaguba** [comment]
I was getting these errors 
```
11: BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
11: BLAS : Bad memory unallocation! :    2  (nil)
```
Googling i found this issue 

https://github.com/autogluon/autogluon/issues/1020

See message
```
This worked for me.
"BLAS stands for Basic Linear Algebra Subprograms. BLAS provides standard interfaces for linear algebra, including BLAS1 (vector-vector operations), BLAS2 (matrix-vector operations), and BLAS3 (matrix-matrix operations).
As per the documentation, if your application is already multi-threaded, it will conflict with OpenBLAS multi-threading. Therefore, you must set OpenBLAS to use a single thread.
So, it seems that your application is conflicting with OpenBLAS multi-threading. You need to run the followings on the command line and it should fix the error:"

export OPENBLAS_NUM_THREADS=1
export GOTO_NUM_THREADS=1
export OMP_NUM_THREADS=1

```
though i did not use the last command, so land is still on 7 threads.

**Comment by ambrad** [comment]
Re: the BLAS error, there is an OLCF issue for it. I believe it's intermittent. I think I had tried the env settings you tried and found that eventually I still got the error. But I might be wrong.

**Review by ambrad — APPROVED** [review]

**Comment by AaronDonahue** [comment]
@oksanaguba , can you check that the labels I added to the PR are correct?  If so, and if this is ready to merge can you add the AUTOMERGE label?

**Comment by oksanaguba** [comment]
@AaronDonahue it is not ready. I have a 512 nodes run in the q with output to check pio issue (which is just to check whether the issue is still there). Then I also want to rerun without io to make sure perf is comparable with Andrew's runs (if perf is degraded, i would need to figure out why before this PR is merged).

Separately, I found slack conversation about BLAS and Andrew was saying there that env settings that are in this PR did not help. @sarats made an OLCF ticket for the issue (not sure how to access OLCF tickets).

**Comment by oksanaguba** [comment]
One more thing -- @AaronDonahue why did you add labels "do not test"? This should be tested at least once, maybe after Frontier runs.

**Comment by bartgol** [comment]
> One more thing -- @AaronDonahue why did you add labels "do not test"? This should be tested at least once, maybe after Frontier runs.

I suppose because none of the mods would affect the AT runs. Of course, someone should check manually on frontier that things work, but there's no point in running tests on blake/weaver/mappy...

**Comment by oksanaguba** [comment]
Thanks to Steve Abbott, who explained to us how cmake in its tests is pulling wrong (not omp) libsci library leading to wrong linking line for the exec. Steve suggested we use a different cc compiler, to try to avoid the issue. After switching compilers, i ran ne30 successfully 14 times, without seeing blas messages in logs. 
```
[onguba@login05.frontier run]$ pwd
/ccs/home/onguba/eff/fgpu-build-apr18b-r32/run
[onguba@login05.frontier run]$ zgrep -i blas *log* | wc -l
0
```
I then submitted a 500-node 2-hour run with the output (this is to test PIO issue)
```
[onguba@login05.frontier run]$ pwd
/ccs/home/onguba/eff/fgpu-build-apr18-r4096/run
[onguba@login05.frontier run]$ ls -la *nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 03:39 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-03000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 03:45 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-06000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 03:51 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-09000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 03:57 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-12000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:03 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-15000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:09 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-18000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:15 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-21000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:21 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-24000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:27 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-27000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:33 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-30000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:38 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-33000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:44 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-36000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:51 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-39000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 04:57 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-42000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 05:02 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-45000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 05:08 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-48000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 05:14 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-51000.nc
-rwxr-xr-t 1 onguba cli115   84758648224 Apr 19 05:21 fgpu-build-apr18-r4096.cpl.hi.0001-01-01-54000.nc
-rwxr-xr-t 1 onguba cli115 7263863461048 Apr 19 05:23 fgpu-build-apr18-r4096.scream.hi.INSTANT.nsteps_x30.0001-01-01-00000.nc
```
Is this the correct output to test FS and IO or is there some other output I need to test?

The I submitted a 500-node run without IO to potentially compare with Andrew's perf runs and that one crashed in the 2nd day with what seems to be a blas issue, but wihtout blas errors:
```
[onguba@login05.frontier run]$ pwd
/ccs/home/onguba/eff/fgpu-build-apr19-r4096/run
[onguba@login05.frontier run]$ grep -i blas *log* | wc -l
0
```
with stack
```
(gdb) bt
#0  0x0000000001524350 in banddiagonal$banddiagonalmod_ ()
#1  0x0000000000d24a42 in solvetemperature$soiltemperaturemod_ ()
#2  0x0000000000d19459 in soiltemperature$soiltemperaturemod_ ()
#3  0x000000000067477b in elm_drv$elm_driver.cray$mt$p0005 ()
#4  0x00007fffe273ff77 in _cray$mt_start_one_code_parallel () from /opt/cray/pe/cce/15.0.0/cce/x86_64/lib/libcraymp.so.1
```

A 500-node  run without lnd threads is in the q.

**Comment by oksanaguba** [comment]
How this setup performs on Frontier for ne1024 runs (repeating some of the info from above):

The issue around land threaded blas was due to cmake not picking up threaded setup during initial tests and then propagating non-threaded linking flags to the e3sm build. The issue was partially resolved by S. Abbott when he suggested we use a different CC compiler (mpicc instead of cc, plus get rid of homp flags, commit 8fcd12e). With ne30 threaded runs, i did not see any more fails (folder fgpu-build-apr18b-r32), and linking flags had no libsci flags.

I then tried to run ne1024 on 512 nodes (fgpu-build-apr19-r4096). It failed in land, without blas memory alloc messages, but nevertheless before land calls its blas tridiag solver. I switched to nontheraed runs (fgpu-build-apr20-r4096), but there was the same issue and the same stack in core.

Andrew suggested a few things -- trying a different node count, 1024 instead of 512, 1 thread, folder fgpu-build-apr26-r8192. It crashed in the same way with the same stack.

Trying debug run on 512 nodes -- folder fgpu-build-apr26-r4096-debug, 1 thread, crashed after 1 hour in land init.

Finally, Andrew found a raw implementation of blas calls for land, which (up to some hiccups, maybe, due to machine issues) allowed me to have a 3-day run, folder  fgpu-build-apr27-r4096-replaceblas .

Performance summary is in the next message.

**Comment by oksanaguba** [comment]
This PR uses rocm5.4, so, comparing with Andrew's run https://github.com/E3SM-Project/perf-data/blob/main/screamv1-frontier-feb2023/data/frontier-v1-scaling1-rocm54-nnodes512.ne1024pg2_ne1024pg2.F2010-SCREAMv1-timing.1272590.230213-015848-model_timing_stats , which (from counts) ran for 0.5 day. Thus my run is 6x longer:

Also, my runs have 
```
namelist_scream.xml:    <disable_diagnostics type="logical">False</disable_diagnostics>
```
which adds a few seconds to the homme timer.

My run 
```
"a:EAMxx::EAMxx::run"                                      -       4096     4096 1.061683e+07   1.764655e+07  4439.111 (  3057      0)  4171.363 (  1721      0)
"a:EAMxx::homme::run"                                      -       4096     4096 1.061683e+07   1.503542e+07  3671.950 (    55      0)  3669.126 (  2604      0)
"a:tl-s prim_step_flexible"                                -       4096     4096 2.123366e+07   1.469887e+07  3589.641 (    82      0)  3587.694 (   737      0)
"a:tl-ae prim_advance_exp"                                 -       4096     4096 1.274020e+08   1.222663e+07  2995.679 (  2971      0)  2978.009 (  2730      0)
"a:ttype9_imex_timestep"                                   -       4096     4096 1.274020e+08   1.052156e+07  2595.286 (  1506      0)  2553.975 (    13      0)

"a:caar compute"                                           -       4096     4096 1.274020e+09   4.765021e+06  1209.973 (  1506      0)  1158.011 (  2914      0)
"a:caar_bexchV"                                            -       4096     4096 6.370099e+08   1.752302e+06   452.999 (  1504      0)   381.518 (  3821      0)
"a:caar limiter"                                           -       4096     4096 7.644119e+08   9.951426e+04    24.543 (  3819      0)    24.021 (  1583      0)
"a:compute_stage_value_dirk"                               -       4096     4096 6.370099e+08   3.815664e+06   957.212 (   236      0)   927.073 (  1751      0)

"a:tl-ae advance_hypervis_dp"                              -       4096     4096 1.274020e+08   1.699394e+06   434.357 (   224      0)   394.151 (  1506      0)
"a:hvf-bhwk"                                               -       4096     4096 1.274020e+08   9.602451e+05   242.650 (   219      0)   228.165 (  1545      0)
"a:hvf-bexch"                                              -       4096     4096 3.822060e+08   6.866599e+05   187.219 (   224      0)   146.706 (  1506      0)

"a:Remap Compute Remap Functor"                            -       4096     4096 6.370099e+07   8.432065e+05   206.434 (  2527      0)   205.011 (  1506      0)

"a:tl-at prim_advec_tracers_compose"                       -       4096     4096 2.123366e+07   1.195408e+06   298.617 (  2730      0)   281.663 (  2971      0)
"a:compose_transport"                                      -       4096     4096 2.123366e+07   1.195166e+06   298.558 (  2730      0)   281.605 (  2971      0)

"a:EAMxx::physics::run"                                    -       4096     4096 1.061683e+07   2.586697e+06   762.780 (  3057      0)   495.199 (  1721      0)
"a:EAMxx::mac_aero_mic::run"                               -       4096     4096 1.061683e+07   7.106872e+05   203.897 (  1899      0)   143.133 (   723      0)
"a:EAMxx::shoc::run"                                       -       4096     4096 1.061683e+07   1.116038e+05    27.619 (  3560      0)    26.663 (  1307      0)
"a:EAMxx::p3::run"                                         -       4096     4096 1.061683e+07   5.379959e+05   161.761 (  1899      0)   101.427 (   723      0)
"a:EAMxx::rrtmgp::run"                                     -       4096     4096 1.061683e+07   1.875846e+06   570.175 (  2848      0)   345.698 (  1721      0)
```

Andrew's
```
"a:EAMxx::EAMxx::run"                                      -       4096     4096 1.769472e+06   2.826695e+06   714.429 (   965      0)   667.167 (  2229      0)
"a:EAMxx::Dynamics::run"                                   -       4096     4096 1.769472e+06   2.393707e+06   584.932 (    56      0)   584.063 (  2605      0)
"a:tl-s prim_step_flexible"                                -       4096     4096 3.538944e+06   2.339856e+06   571.636 (   646      0)   571.030 (  3528      0)
"a:tl-ae prim_advance_exp"                                 -       4096     4096 2.123366e+07   1.935615e+06   475.258 (  3686      0)   470.038 (  2392      0)
"a:ttype9_imex_timestep"                                   -       4096     4096 2.123366e+07   1.669444e+06   411.608 (  3821      0)   404.324 (  2150      0)

"a:caar compute"                                           -       4096     4096 2.123366e+08   7.458267e+05   183.501 (  3992      0)   181.234 (  1950      0)
"a:caar_bexchV"                                            -       4096     4096 1.061683e+08   2.569460e+05    66.181 (  1188      0)    58.478 (  3520      0)
"a:caar limiter"                                           -       4096     4096 1.274020e+08   1.614335e+04     4.006 (  2438      0)     3.900 (  2507      0)
"a:compute_stage_value_dirk"                               -       4096     4096 1.061683e+08   6.359027e+05   160.427 (  1210      0)   154.301 (  3702      0)

"a:tl-ae advance_hypervis_dp"                              -       4096     4096 2.123366e+07   2.659433e+05    69.910 (   395      0)    62.318 (  1768      0)
"a:hvf-bhwk"                                               -       4096     4096 2.123366e+07   1.543550e+05    42.184 (  1983      0)    36.576 (  3644      0)
"a:hvf-bexch"                                              -       4096     4096 6.370099e+07   9.757230e+04    28.783 (   392      0)    21.096 (  1768      0)

"a:Remap Compute Remap Functor"                            -       4096     4096 1.061683e+07   1.405079e+05    34.473 (  2855      0)    34.129 (  2826      0)

"a:tl-at prim_advec_tracers_compose"                       -       4096     4096 3.538944e+06   1.920946e+05    49.355 (  2703      0)    44.207 (  3686      0)
"a:compose_transport"                                      -       4096     4096 3.538944e+06   1.920688e+05    49.349 (  2703      0)    44.197 (  3686      0)

"a:EAMxx::physics::run"                                    -       4096     4096 1.769472e+06   4.288214e+05   128.924 (   965      0)    81.889 (  2229      0)
"a:EAMxx::mac_aero_mic::run"                               -       4096     4096 1.769472e+06   1.166641e+05    35.216 (  3398      0)    23.445 (  2481      0)
"a:EAMxx::Macrophysics::run"                               -       4096     4096 1.769472e+06   1.838574e+04     4.575 (  2760      0)     4.398 (  3029      0)
"a:EAMxx::Microphysics::run"                               -       4096     4096 1.769472e+06   8.880932e+04    28.377 (  3398      0)    16.779 (  2481      0)
"a:EAMxx::Radiation::run"                                  -       4096     4096 1.769472e+06   3.121216e+05    94.780 (  3232      0)    57.484 (  1863      0)
```

**Comment by oksanaguba** [comment]
Comparing with Andrews run
```
Physics::run -- 762/6=127 (mine) compared to AMB 129

Homme 3671/6=610 compared to 585

Caar compute 1209/6=201 compared to 184

Can exchange 453/6=76 compared to 66

Dirk 957/6=160 compared to 160

advance_hv_dp 434/6=72 compared to 70

Compose 299/6=50 compared to 49
```

The loss is due to diagnostics on, but there is significant change in caar and caar exchange. Not sure yet if we can account for the changes. There were barriers inserted in caar, and also new HV profiles added some code to HV parts.

**Comment by oksanaguba** [comment]
Also, i ran this branch with output (comment above https://github.com/E3SM-Project/scream/pull/2279#issuecomment-1516727673 ) and did not experience any crashes. Though I am not sure I ran with the output that needs to be tested.

**Comment by ambrad** [comment]
The performance diffs are fine. My runs were in mid Feb, with master 7 Feb; the exact commit history is here: https://github.com/E3SM-Project/scream/commits/archive/screamv1-frontier-feb2023. As Oksana wrote, the only mystery is the "CAAR compute" timer. Boundary exchange timers are highly variable, and the other timers are all about the same. I consider the "CAAR compute" timer mystery to have too low a priority to investigate.

**Comment by ambrad** [comment]
For the ELM change, to avoid issues when merging SCREAM into upstream (which we do now on occasion), I recommend the following steps:

1. Separate the ELM change into a separate PR.
2. Merge this PR.
3. Create a branch called "machines/frontier".
4. Add the ELM change to it and push the branch. This branch will then be the branch on which we do production runs on Frontier. We'll merge master into it periodically. We'll do this same type of thing for every production machine that has things we have to work around.

File: components/elm/src/biogeophys/BandDiagonalMod.F90 [hunk 1]
```diff
@@ -218,4 +218,1890 @@ subroutine BandDiagonal(bounds, lbj, ubj, jtop, jbot, numf, filter, nband, b, r,
 
   end subroutine BandDiagonal
 
+
+
+
+
+      SUBROUTINE DCOPY(N,DX,INCX,DY,INCY)
```
Comment (ambrad) [inline components/elm/src/biogeophys/BandDiagonalMod.F90:225]: Preface this code with a comment about its source (the URL I gave you) and a link to this license: https://netlib.org/lapack/LICENSE.txt

**Comment by oksanaguba** [comment]
I could not run a test suite the way i usually run due to some python error, so I ran instead
```
~/scream-master/cime/scripts/create_test ERS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.crusher-scream-gpu_crayclang-scream
```
without baselines and got 2 sets of comparisons pass
```
tail -n20 /lustre/orion/cli133/proj-shared/onguba/e3sm_scratch/crusher/ERS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.crusher-scream-gpu_crayclang-scream.20230507_000357_nqkdsg/run/ERS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.crusher-scream-gpu_crayclang-scream.20230507_000357_nqkdsg.scream.hi.AVERAGE.nsteps_x362.0001-01-01-00000.nc.base.cprnc.out

...

tail -n20 /lustre/orion/cli133/proj-shared/onguba/e3sm_scratch/crusher/ERS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.crusher-scream-gpu_crayclang-scream.20230507_000357_nqkdsg/run/ERS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.crusher-scream-gpu_crayclang-scream.20230507_000357_nqkdsg.cpl.hi.0001-01-08-46800.nc.base.cprnc.out
```

**Comment by oksanaguba** [comment]
@ambrad @AaronDonahue  i made a branch with blas impl last week https://github.com/E3SM-Project/scream/compare/master...oksanaguba/machines/frontier and added info on ref implementation, too. i am running a bunch of runs from this branch to create a ticket for OLCF about crashes, but this branch can be merged. i will remove land changes now and it can be then merged. thanks!

**Review by ambrad — APPROVED** [review]

**Review by ambrad — APPROVED** [review]

**Comment by oksanaguba** [comment]
also i ran e3sm_scream_v1 suite as Andrew suggested, no fails, 
```
[onguba@login09.frontier scream-b3]$ cd -
/lustre/orion/cli133/proj-shared/onguba/e3sm_scratch/crusher
[onguba@login09.frontier crusher]$ ./cs.status.20230508_113423_xyx67m | grep -i fail
[onguba@login09.frontier crusher]$ 
```

**Pull request merged**